llvm/lib/Analysis/HashRecognize.cpp

```diff
@@ -185,9 +199,14 @@ KnownBits ValueEvolution::computeInstr(const Instruction *I) {
   // Compute the KnownBits for a Select(Cmp()), forcing it to take the branch
   // that is predicated on the (least|most)-significant-bit check.
   CmpPredicate Pred;
-  Value *L, *R, *TV, *FV;
-  if (match(I, m_Select(m_ICmp(Pred, m_Value(L), m_Value(R)), m_Value(TV),
-                        m_Value(FV)))) {
+  Value *L, *R;
+  Instruction *TV, *FV;
+  if (match(I, m_Select(m_ICmp(Pred, m_Value(L), m_Value(R)), m_Instruction(TV),
+                        m_Instruction(FV)))) {
+    Visited.insert(cast<Instruction>(I->getOperand(0)));
+    Visited.insert(TV);
+    Visited.insert(FV);
```

Comment:
Shouldn't we rather rely on `compute(TV)` and `compute(FV)` below?

Reply:
Thanks for the comment, but it should be close to these `insert`s. It is risky to do the inserts without any checks.
Can we have `insert(FV);` before `compute(TV);` and preferrably some `assert` on `FV` ? It is not clear that `isBigEndianBitShift` checks it. Similarly `insert(TV);` before `compute(FV);`.

Reply:
I plan to write thorough documentation for HashRecognize after we get the optimization enabled: in retrospect, several of the things I wrote down as comments are not evident from the code.